frame/bridge/relayer-game/src/lib.rs

```diff
@@ -0,0 +1,548 @@
+		/// Each target chain's header relay can open a game
+		pub Proposals
+			get(fn proposals_of_game)
+			: map hasher(blake2_128_concat) GameId<TcBlockNumber<T, I>>
+			=> Vec<Proposal<
+				AccountId<T>,
+				RingBalance<T, I>,
+				TcHeaderId<TcBlockNumber<T, I>, TcHeaderHash<T, I>>,
+				ProposalId<TcBlockNumber<T, I>, TcHeaderHash<T, I>>
+			>>;
+
+		/// The allow samples for each game
+		pub Samples
+			get(fn samples_of_game)
+			: map hasher(blake2_128_concat) TcBlockNumber<T, I>
+			=> Vec<TcBlockNumber<T, I>>;
+
+		/// The closed rounds which had passed the challenge time at this moment
+		pub ClosedRounds
+			get(fn closed_rounds_at)
+			: map hasher(blake2_128_concat) BlockNumber<T>
+			=> Vec<(GameId<TcBlockNumber<T, I>>, Round)>;
+
+		/// All the `TcHeader`s store here, **NON-DUPLICATIVE**
+		pub TcHeaders
+			get(fn tc_header)
+			: map hasher(blake2_128_concat) TcHeaderId<TcBlockNumber<T, I>, TcHeaderHash<T, I>>
+			=> RefTcHeader;
+
+		/// The finalize blocks' header's id which is recorded in darwinia
+		pub ConfirmedTcHeaderIds
+			get(fn confirmed_tc_header_id)
+			: TcHeaderId<TcBlockNumber<T, I>, TcHeaderHash<T, I>>;
+
+		pub Bonds
+			get(fn bond_of_relayer)
+			: map hasher(blake2_128_concat) AccountId<T>
+			=> RingBalance<T, I>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait<I>, I: Instance = DefaultInstance> for enum Call
+	where
+		origin: T::Origin
+	{
+		type Error = Error<T, I>;
+
+		fn deposit_event() = default;
+
+		// TODO: too many db operations? move to `offchain_worker`?
+		fn on_finalize(block_number: BlockNumber<T>) {
+			let proposals = <ClosedRounds<T, I>>::take(block_number);
+
+			if proposals.len() == 0 {
+				return;
+			}
+
+			for (game_id, round) in proposals {
+				let proposals = Self::proposals_of_game(game_id)
+					.into_iter()
+					.filter(|proposal|
+						T::RelayerGameAdjustor::round_from_chain_len(proposal.chain.len() as _)
+							== round
+					)
+					.collect::<Vec<_>>();
+				let proposals_len = proposals.len();
+
+				if proposals_len == 0 {
+					return;
+				}
+
+				if proposals_len == 1 {
+					// Chain's len is ALWAYS great than 1 under this match pattern; qed
+					let proposal = proposals[0].clone();
+					let mut extend_from = proposal.extend_from.clone();
+
+					while let Some((extend_from_block_number, extend_from_header_hash))
+						= extend_from.clone()
+					{
+						let mut relayer = None;
+						let mut evils = vec![];
+
+						for proposal in <Proposals<T, I>>::mutate(
+							extend_from_block_number,
+							|proposals| proposals
+								.drain_filter(|proposal|
+									T::RelayerGameAdjustor::round_from_chain_len(
+										proposal.chain.len() as _
+									) != round
+								)
+								.collect::<Vec<_>>()
+						) {
+							if let Some(BondedTcHeader { id, bond }) = proposal.chain.last() {
+								let (_, header_hash) = id;
+								let mut header = <TcHeaders<T, I>>::take(id);
```

Comment:
There are two write operation with`take` then `insert` operations.  
Maybe we should try with `get` then `kill` operations with less write operation.

Reply:
before: 
if 0 ref_count: 1w
if x ref_count: 2w
after:
if 0 ref_count: 1r 1w
if x ref_count: 1r 1w

I'm not sure how to deal with this

Reply:
Ref removed.